ttnn/cpp/ttnn/device_operation.hpp

```diff
@@ -0,0 +1,294 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <concepts>
+#include <optional>
+#include <tt_eager/tensor/tensor.hpp>
+
+#include "third_party/magic_enum/magic_enum.hpp"
+#include "tt_dnn/op_library/operation_history.hpp"
+#include "tt_stl/concepts.hpp"
+#include "tt_stl/reflection.hpp"
+#include "tt_stl/unique_any.hpp"
+
+namespace ttnn {
+
+namespace device_operation {
+
+template <typename... program_attributes_t>
+struct CachedProgram {
+    tt::tt_metal::Program program;
+    // Cached program needs to share program_attributes between create and override_runtime_arguments functions
+    std::tuple<program_attributes_t...> program_attributes;
+
+    CachedProgram(tt::tt_metal::Program&& program, program_attributes_t... program_attributes) :
+        program{std::move(program)}, program_attributes{std::tuple{program_attributes...}} {}
+};
+
+struct ProgramCacheItem {
```

Comment:
Can this be generic so that it can cache a sequence of {enqueu programs, allocates, deallocates , ...} any trace of Metal run time commands?